eeps/eep-0066.md

```diff
@@ -0,0 +1,340 @@
+have to be interpreted as prefixed string start «`b<`» with «`X`»
+being the first string content character.
+
+For the «`/`» character we run into similar problems with for example
+«`b/X`», which would be a run-time error today, but if we also would
+want capital letter prefixes, then «`B/X`» is perfectly valid today
+but would become a string start.
+
+There are more likely problems with simple string prefixes:
+«`#bf{`» is today the start of a record named «`bf`», and is
+scanned as punctuation character «`#`», atom «`bf`» and separator «`{`»,
+which the parser sorts out to be a record start.
+
+With simple prefix characters the tokenizer would have to be rewritten
+to recognize «`#bf`» as a new record token, a rewrite that might cause
+unexpected changes in record handling.  For example, today, «`# bf {`»
+is also a valid record start, so to be completely compatible the tokenizer
+would have to allow white-space or even newlines within the new record
+token, between «`#`» and the atom characters, which would be really ugly...
+
+For other reasons, namely that function call parenthesis are optional,
+Elixir has chosen to use the «`~`» character as the start of
+a string prefix which they call a "[Sigil][1]".
+
+Having a distinct start character for this feature simplifies
+tokenizing and parsing.
+
+### Sigil
+
+In a general sense, a [Sigil][3], is a prefix to a variable
+that indicates its type, such as «`$I`» in Basic or Perl,
+where «`$`» is the sigil and «`I`» is the variable.
+
+Here we define a Sigil as a prefix (and a suffix) to a string literal
+that indicates how it should be interpreted.  The Sigil is
+a syntactic sugar that creates some Erlang term.
+
+A Sigil string literal consists of:
+
+1. The Sigil start character «`~`».
+2. The [Sigil Type][], a name that may be empty.
+3. The [String Content][] within [String Delimiters][].
+4. The [Sigil Suffix][], a name character sequence that may be empty.
+
+### Sigil Type
+
+Immediately following the [Sigil][] start character is the Sigil Type
+which is a name composed of a sequence of characters that are allowed
+as the second or later characters in a variable or an atom.
+In short ISO [Latin-1][] letters, digits, «`_`» and «`@`».
+The [Sigil Type][] may be empty.
+
+The Sigil Type defines how the [Sigil][] syntactic sugar
+shall be interpreted.  The suggested Sigil Types are:
+
+* «»: the Vanilla [Sigil][].
+
+  Creates a `unicode:unicode_binary()` in Erlang.
+  It is a string represented as a UTF-8 encoded binary,
+  equivalent to applying `unicode:characters_to_binary/1`
+  on the [String Content][].  The [String Delimiters][]
+  and escape characters work as for regular strings,
+  triple-quoted strings, or quoted atoms in Erlang.
+
+  So «`~"abc\d"`» is equivalent to «`<<"abc\d"/utf8>>`», and
+  «`~'abc"d'`» is equivalent to «`<<"abc\"d"/utf8>>`».
+
+* «`r`»: regular expression.
+
+  Creates a term `{RE::unicode:charlist(),Flags::[unicode:latin1_char()]}`
```

Comment:
Perhaps the topic of another conversation, but Elixir compiles the regular expression at compilation time. We have to store the endianess and re version, and recompile on the fly if those change, but Erlang could consider doing the same.

Reply:
I will add a chapter about the reasoning behind this sigil output type.  When to compile an RE is part of that.

So far, to my knowledge, we don't have anything cached for a specific platform in the .beam files.  We have talked about compiling an RE at load time.  That is also when the Just In Time compiler runs.

Reply:
I don't think compiling REs at load-time is a good idea: if we abandon PCRE we'd start getting mysterious failures at load-time that may be difficult for the user to pinpoint. `on_load` also has nasty implications for code loading which, among other things, prevent loading modules in parallel.

I think it's better to leave it to the user, there's little stopping them from compiling the REs in an initialization phase and putting them in `persistent_term`.